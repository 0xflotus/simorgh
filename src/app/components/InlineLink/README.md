# Inline Link [![Storybook](https://github.com/storybooks/press/blob/master/badges/storybook.svg)](https://simorghstorybook.now.sh/?selectedKind=InlineLink)

## Description

The `Inline Link` component is made up of a styled HTML `<a>` element with an `href` attribute. Links open in the current tab. 

## When to use this component

The purpose of this component is to return an inline link element.

`Inline Link` is made use of indirectly in the [Text](../Text) component, which uses the [Inline Link Container](../../containers/InlineLink) to render an `Inline Link` differently based on whether the href points in or outisde of Simorgh.

The [Paragraph Container](../../containers/Paragraph) and the [Inline Link Container](../../containers/InlineLink) also render the `Inline Link`.

## Accessibility notes 

The `Inline Link` includes:
* Generous tap target across breakpoints, for ease of tapping
* Link states (default, hover, focus and visited) that meet contrast ratios specified in [BBC Future Media Standards & Guidelines](http://www.bbc.co.uk/guidelines/futuremedia/accessibility/html/colour-contrast.shtml) for accessibility
* Keyboard operability. Users can navigate to and operate each `Inline Link` using the keyboard alone. This in turn enables `Inline Link`'s operability by screen readers, alternative keyboards or input devices.
