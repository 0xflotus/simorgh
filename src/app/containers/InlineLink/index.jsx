import React from 'react';
import { Link } from 'react-router-dom';
import { string, node } from 'prop-types';
import InlineLink from '../../components/InlineLink';

const InternalInlineLink = InlineLink.withComponent(Link);

const InlineLinkContainer = ({ href, children, ...rest }) => {
  // if URL matches a valid route, use a react-router link
  if (href.match('^/news/articles/c0{8}[0-9]{2}o$')) {
    return (
      <InternalInlineLink to={href} {...rest}>
        {children}
      </InternalInlineLink>
    );
  }

  // else return a normal hyperlink
  return (
    <InlineLink href={href} {...rest}>
      {children}
    </InlineLink>
  );
};

InlineLinkContainer.propTypes = {
  href: string.isRequired,
  children: node.isRequired,
};

export default InlineLinkContainer;
