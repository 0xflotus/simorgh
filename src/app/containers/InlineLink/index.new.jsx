import React from 'react';
import { Link } from 'react-router-dom';
import InlineLink from '../../components/InlineLink';
import Blocks from '../Blocks';
import fragment from '../Fragment';
import { inlineLinkModelPropTypes } from '../../models/propTypes/inlineLink';

const InternalInlineLink = InlineLink.withComponent(Link);

const componentsToRender = { fragment };

const InlineLinkContainer = ({ locator, blocks }) => {
  const regex = '^/news/articles/c[a-zA-Z0-9]{10}o$';

  // if URL matches a valid route, use a react-router link
  if (locator.match(regex)) {
    return (
      <InternalInlineLink to={locator}>
        <Blocks blocks={blocks} componentsToRender={componentsToRender} />
      </InternalInlineLink>
    );
  }

  // else return a normal hyperlink
  return (
    <InlineLink href={locator}>
      <Blocks blocks={blocks} componentsToRender={componentsToRender} />
    </InlineLink>
  );
};

InlineLinkContainer.propTypes = inlineLinkModelPropTypes;

export default InlineLinkContainer;
