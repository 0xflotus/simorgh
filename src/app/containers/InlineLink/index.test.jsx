import React from 'react';
import { StaticRouter } from 'react-router-dom';
import {
  shouldMatchSnapshot,
  shouldShallowMatchSnapshot,
} from '../../helpers/tests/testHelpers';
import InlineLink from './index';

describe('InlineLink', () => {
  describe('with link matching routes for SPA', () => {
    shouldMatchSnapshot(
      'should render correctly',
      /*
        for the value it would bring, it is much simpler to wrap a react-router Link in a Router, rather than mock a Router or pass come mocked context.
      */
      <StaticRouter>
        <InlineLink href="https://www.bbc.co.uk/news/articles/c0000000027o">
          Internal SPA article
        </InlineLink>
      </StaticRouter>,
    );

    shouldMatchSnapshot(
      'should render correctly for TEST',
      /*
        for the value it would bring, it is much simpler to wrap a react-router Link in a Router, rather than mock a Router or pass come mocked context.
      */
      <StaticRouter>
        <InlineLink href="https://www.test.bbc.co.uk/news/articles/c0000000027o">
          Internal SPA article
        </InlineLink>
      </StaticRouter>,
    );
  });

  describe('with link not matching SPA route', () => {
    shouldShallowMatchSnapshot(
      'should render correctly',
      <InlineLink href="https://www.bbc.com/news">BBC News</InlineLink>,
    );
  });
});
