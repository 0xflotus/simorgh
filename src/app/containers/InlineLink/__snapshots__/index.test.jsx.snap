// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineLinkContainer external link accessibility should be explicitly marked "external" for screen reader users & localised 1`] = `
<Context.Provider
  value={
    Object {
      "articleAuthor": "https://www.facebook.com/bbcnews",
      "brandName": "BBC News فارسی",
      "defaultImage": "https://news.files.bbci.co.uk/ws/img/logos/og/persian.png",
      "defaultImageAltText": "BBC News فارسی",
      "externalLinkText": "، لینک خارجی",
      "imageCaptionOffscreenText": " ، عنوان تصویر",
      "imageCopyrightOffscreenText": " ، منبع تصویر",
      "locale": "fa",
      "service": "persian",
      "serviceName": "Persian",
      "themeColor": "#B80000",
      "translations": Object {
        "error": Object {
          "404": Object {
            "callToActionFirst": "روش دیگر، لطفآ به ",
            "callToActionLast": " بازدید کنید",
            "callToActionLinkText": "صفحه اصلی بی بی سی فارسی",
            "callToActionLinkUrl": "https://www.bbc.com/persian",
            "message": "متاسفانه به دلیلی این صفحه در دسترس نمی‌باشد. لطفآ این رو امتحان کنید:",
            "solutions": Array [
              "لینک را چک کنید",
              "صفحه را ریفرش کنید",
              "با استفاده از نوار جستجوی بی بی سی برای این صفحه جستجو کنید",
            ],
            "statusCode": "۴۰۴",
            "title": "این صفحه یافت نشد",
          },
          "500": Object {
            "callToActionFirst": "روش دیگر، لطفآ به ",
            "callToActionLast": " بازدید کنید",
            "callToActionLinkText": "صفحه اصلی بی بی سی فارسی",
            "callToActionLinkUrl": "https://www.bbc.com/persian",
            "message": "متاسفانه به دلیلی این صفحه در دسترس نمی‌باشد. لطفآ این رو امتحان کنید:",
            "solutions": Array [
              "صفحه را ریفرش کنید",
              "بعدا برگردید",
            ],
            "statusCode": "۵۰۰",
            "title": "مشکلی پیش آمده است",
          },
        },
      },
      "twitterCreator": "@bbcpersian",
      "twitterSite": "@bbcpersian",
    }
  }
>
  <InlineLinkContainer
    blocks={
      Array [
        Object {
          "model": Object {
            "attributes": Array [
              "",
            ],
            "text": "این لینک هست",
          },
          "type": "fragment",
        },
      ]
    }
    isExternal={true}
    locator="https://www.example.com/"
  />
</Context.Provider>
`;

exports[`InlineLinkContainer external link accessibility should be explicitly marked "external" for screen reader users 1`] = `
<Context.Provider
  value={
    Object {
      "articleAuthor": "https://www.facebook.com/bbcnews",
      "brandName": "BBC News",
      "defaultImage": "https://www.bbc.co.uk/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1",
      "defaultImageAltText": "BBC News",
      "externalLinkText": ", external",
      "imageCaptionOffscreenText": "Image caption, ",
      "imageCopyrightOffscreenText": "Image source, ",
      "locale": "en_GB",
      "service": "news",
      "serviceName": "News",
      "themeColor": "#B80000",
      "translations": Object {
        "error": Object {
          "404": Object {
            "callToActionFirst": "Alternatively, please visit the ",
            "callToActionLast": "",
            "callToActionLinkText": "BBC News homepage.",
            "callToActionLinkUrl": "https://www.bbc.com/news",
            "message": "Sorry, we're unable to bring you the page you’re looking for. Please try:",
            "solutions": Array [
              "Double checking the url",
              "Hitting the refresh button in your browser",
              "Searching for this page using the BBC search bar",
            ],
            "statusCode": "404",
            "title": "Page cannot be found",
          },
          "500": Object {
            "callToActionFirst": "Alternatively, please visit the ",
            "callToActionLast": "",
            "callToActionLinkText": "BBC News homepage.",
            "callToActionLinkUrl": "https://www.bbc.com/news",
            "message": "Sorry, we’re currently unable to bring you the page you're looking for. Please try:",
            "solutions": Array [
              "Hitting the refresh button in your browser",
              "Coming back again later",
            ],
            "statusCode": "500",
            "title": "Internal server error",
          },
        },
      },
      "twitterCreator": "@BBCNews",
      "twitterSite": "@BBCNews",
    }
  }
>
  <InlineLinkContainer
    blocks={
      Array [
        Object {
          "model": Object {
            "attributes": Array [],
            "text": "This is a link",
          },
          "type": "fragment",
        },
      ]
    }
    isExternal={true}
    locator="https://www.example.com/"
  />
</Context.Provider>
`;

exports[`InlineLinkContainer internal link not matching SPA route should render correctly 1`] = `
<ServiceContextConsumer>
  [Function]
</ServiceContextConsumer>
`;

exports[`InlineLinkContainer link matching routes for SPA should render correctly 1`] = `
.c0 {
  color: #0F556C;
  padding: 2px;
  border-bottom: 1px solid #0F556C;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:visited {
  color: #5C5752;
  border-bottom: 1px solid #5C5752;
}

.c0:focus {
  background-color: #0F556C;
  border-bottom: 1px solid #F5F3F1;
  color: #F5F3F1;
}

.c0:hover {
  color: #0F556C;
  background-color: #C3DEE7;
  border-bottom: 1px solid #0F556C;
}

<a
  className="-Link c0"
  href="/news/articles/c85pqyj5m2ko"
  onClick={[Function]}
>
  This is text for an internal link
</a>
`;

exports[`InlineLinkContainer link matching routes for SPA should render correctly for TEST environment 1`] = `
.c0 {
  color: #0F556C;
  padding: 2px;
  border-bottom: 1px solid #0F556C;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:visited {
  color: #5C5752;
  border-bottom: 1px solid #5C5752;
}

.c0:focus {
  background-color: #0F556C;
  border-bottom: 1px solid #F5F3F1;
  color: #F5F3F1;
}

.c0:hover {
  color: #0F556C;
  background-color: #C3DEE7;
  border-bottom: 1px solid #0F556C;
}

<a
  className="-Link c0"
  href="/news/articles/c85pqyj5m2ko"
  onClick={[Function]}
>
  This is text for an internal link
</a>
`;
